hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieWriteClient.java

```diff
@@ -237,6 +251,30 @@ protected void commit(HoodieTable table, String commitActionType, String instant
     HoodieActiveTimeline activeTimeline = table.getActiveTimeline();
     // Finalize write
     finalizeWrite(table, instantTime, stats);
+    // do save internal schema to support Implicitly add columns in write process
+    if (!metadata.getExtraMetadata().containsKey(SerDeHelper.LATEST_SCHEMA)
+        && metadata.getExtraMetadata().containsKey(SCHEMA_KEY) && table.getConfig().getSchemaEvolutionEnable()) {
+      TableSchemaResolver schemaUtil = new TableSchemaResolver(table.getMetaClient());
+      String historySchemaStr = schemaUtil.getTableHistorySchemaStrFromCommitMetadata().orElse("");
+      FileBasedInternalSchemaStorageManager schemasManager = new FileBasedInternalSchemaStorageManager(table.getMetaClient());
+      if (!historySchemaStr.isEmpty()) {
+        InternalSchema internalSchema = InternalSchemaUtils.searchSchema(Long.parseLong(instantTime),
+            SerDeHelper.parseSchemas(historySchemaStr));
+        Schema avroSchema = HoodieAvroUtils.createHoodieWriteSchema(new Schema.Parser().parse(config.getSchema()));
+        InternalSchema evolutionSchema = AvroSchemaEvolutionUtils.evolveSchemaFromNewAvroSchema(avroSchema, internalSchema);
+        if (evolutionSchema.equals(internalSchema)) {
+          metadata.addMetadata(SerDeHelper.LATEST_SCHEMA, SerDeHelper.toJson(evolutionSchema));
+          schemasManager.persistHistorySchemaStr(instantTime, historySchemaStr);
```

Comment:
If the Schema is unchanged, can we avoid writing the history schema ? We would need to change the InternalSchemaUtils.searchSchema code to look for nearest version id smaller than the query Version to resolve.

Reply:
if schema is unchanged, we just copy the old history schema,  no extra schema info will be added.
of course we can avoid writeing history schema, directly,  let me modfiy those codes

Reply:
I thought about it. Maybe I can't modify it like this

1) Hoodie will archive and clean up expired commit files. If the version of historyschema lags behind the newest commit too much, we will not be able to determine whether these historyschemas are valid

2) When we save the historyschema, we will automatically clean up the number of historyschema. By default, the reserved number is no more than 10, so the number of historyschema files will not become large

Reply:
TODO added.
fixed